narwhal/types/src/primary.rs

```diff
@@ -1183,6 +1225,278 @@ impl CertificateV1 {
+                sigs.iter().map(|(_, sig)| sig).collect(),
+            )
+            .map_err(|_| DagError::InvalidSignature)?
+        };
+
+        let aggregate_signature_bytes = AggregateSignatureBytes::from(&aggregated_signature);
+
+        let aggregate_signature_state = if !check_stake {
+            AggregateSignatureState::Unsigned(aggregate_signature_bytes)
+        } else {
+            AggregateSignatureState::Unverified(aggregate_signature_bytes)
+        };
+
+        Ok(Certificate::V2(CertificateV2 {
+            header,
+            aggregate_signature_state,
+            signed_authorities,
+            metadata: Metadata::default(),
+        }))
+    }
+
+    /// This function requires that certificate was verified against given committee
+    pub fn signed_authorities(&self, committee: &Committee) -> Vec<PublicKey> {
+        assert_eq!(committee.epoch(), self.epoch());
+        let (_stake, pks) = self.signed_by(committee);
+        pks
+    }
+
+    pub fn signed_by(&self, committee: &Committee) -> (Stake, Vec<PublicKey>) {
+        // Ensure the certificate has a quorum.
+        let mut weight = 0;
+
+        let auth_indexes = self.signed_authorities.iter().collect::<Vec<_>>();
+        let mut auth_iter = 0;
+        let pks = committee
+            .authorities()
+            .enumerate()
+            .filter(|(i, authority)| match auth_indexes.get(auth_iter) {
+                Some(index) if *index == *i as u32 => {
+                    weight += authority.stake();
+                    auth_iter += 1;
+                    true
+                }
+                _ => false,
+            })
+            .map(|(_, authority)| authority.protocol_key().clone())
+            .collect();
+        (weight, pks)
+    }
+
+    /// Verifies the validity of the certificate.
+    /// TODO: Output a different type, similar to Sui VerifiedCertificate.
+    pub fn verify(&mut self, committee: &Committee, worker_cache: &WorkerCache) -> DagResult<()> {
+        // Ensure the header is from the correct epoch.
+        ensure!(
+            self.epoch() == committee.epoch(),
+            DagError::InvalidEpoch {
+                expected: committee.epoch(),
+                received: self.epoch()
+            }
+        );
+
+        // Genesis certificates are always valid.
+        if self.round() == 0 && Self::genesis(committee).contains(self) {
+            return Ok(());
+        }
+
+        // Save signature verifications when the header is invalid.
+        self.header.validate(committee, worker_cache)?;
+
+        let (weight, pks) = self.signed_by(committee);
+
+        ensure!(
+            weight >= committee.quorum_threshold(),
+            DagError::CertificateRequiresQuorum
+        );
+
+        let aggregrate_signature_bytes = match self.aggregate_signature_state {
+            AggregateSignatureState::VerifiedViaLeader(ref bytes) => bytes,
+            AggregateSignatureState::VerifiedDirectly(_) => return Ok(()),
```

Comment:
Maybe I would avoid using this `return Ok(())` here as I think is a bit hard to distinct and anyone who would like to add some extra verification let's say in the end of the code could miss that and their check skipped by this. One thing you could do is extracting all the signature verification logic to a separate method `verify_signature` that would be dedicated to this. The method could still return an error in case this would fail or just return immediately as you do here. Ex

```
ensure!(
    weight >= committee.quorum_threshold(),
    DagError::CertificateRequiresQuorum
);
.....

self.verify_signature()?

....
Ok(())
```